clang/lib/CodeGen/CodeGenTypes.cpp

```diff
@@ -51,6 +51,55 @@ void CodeGenTypes::addRecordTypeName(const RecordDecl *RD,
                                      StringRef suffix) {
   SmallString<256> TypeName;
   llvm::raw_svector_ostream OS(TypeName);
+  // If RD is spirv_JointMatrixINTEL type, mangle differently.
+  if (CGM.getTriple().isSPIRV() || CGM.getTriple().isSPIR()) {
+    if (RD->getQualifiedNameAsString() == "__spv::__spirv_JointMatrixINTEL") {
+      if (auto TemplateDecl = dyn_cast<ClassTemplateSpecializationDecl>(RD)) {
+        auto TemplateArgs = TemplateDecl->getTemplateArgs().asArray();
+        OS << "spirv.JointMatrixINTEL.";
+        for (auto &TemplateArg : TemplateArgs) {
+          OS << "_";
+          if (TemplateArg.getKind() == TemplateArgument::Type) {
+            llvm::Type *TTy = ConvertType(TemplateArg.getAsType());
+            if (TTy->isIntegerTy()) {
+              switch (TTy->getIntegerBitWidth()) {
+              case 8:
+                OS << "char";
+                break;
+              case 16:
+                OS << "short";
+                break;
+              case 32:
+                OS << "int";
+                break;
+              case 64:
+                OS << "long";
+                break;
+              default:
+                OS << "i" << TTy->getIntegerBitWidth();
+                break;
+              }
+            } else if (TTy->isBFloatTy()) {
+              OS << "bfloat16";
+            } else if (TTy->isStructTy()) {
+              StringRef LlvmTyName = TTy->getStructName();
+              // Emit half/bfloat16 for cl::sycl[::*]::{half,bfloat16}
+              if (LlvmTyName.startswith("class.cl::sycl::") ||
+                  LlvmTyName.startswith("class.__sycl_internal::"))
```

Comment:
FYI: I think these are being re-named: `cl::sycl` -> `sycl`.
One of @aelovikov-intel commits removing `cl::` namespace will conflict with this change (e.g. https://github.com/intel/llvm/pull/6518).

Reply:
That's true, but I think it should be resolved in the PR that would be committed last, not first.

Reply:
I'll keep an eye on that PR, and if that lands before this, I'll fix it here.